### DISCUSSION TITLE
Unexpected simulation time extension when defining a multilevel atomic susceptibility structure with more blocks

### DISCUSSION THREAD

**Discussion by haohe0417** [NanoComp/meep#2417] [Q&A] [answered]
Hi all,

I am simulating a laser system using multilevel-atomic susceptibility model. I need to initialize the atomic population inhomogeneously over space, but the `initial_populations` in the `MultilevelAtom` class cannot be assigned as a function of spatial coordinates. So I divided the multilevel atomic structure into several blocks. Each block has a different value of `initial_populations`. However, the simulation becomes very time-consuming as I do so.

To show this issue, we take the original tutorial script [`multilevel-atom.py`](https://github.com/NanoComp/meep/blob/master/python/examples/multilevel-atom.py) as an example. 

In that example, the simulation only takes ~200s. Then I divided the single two-level material block into 10 blocks with different `initial_populations`. Below are the codes, 

````python
geometry = []
for i in range(10):
    transitions = [
        mp.Transition(
            1,
            2,
            pumping_rate=Rp,
            frequency=freq_21,
            gamma=gamma_21,
            sigma_diag=mp.Vector3(sigma_21, 0, 0),
        ),
        mp.Transition(2, 1, transition_rate=rate_21),
    ]
    ml_atom = mp.MultilevelAtom(sigma=1, transitions=transitions, initial_populations=[N0*(i/10)])
    two_level = mp.Medium(index=ncav, E_susceptibilities=[ml_atom])
    # Specify the cavity geometry:
    geometry.append(
        mp.Block(
            center=mp.Vector3(z=-0.5 * sz + (i+1/2) * (Lcav/10)),
            size=mp.Vector3(mp.inf, mp.inf, (Lcav/10)),
            material=two_level,
        ))
````

the simulation time raised from ~200s to ~1440s. It can even go up to ~50000s if I define the material pixel by pixel(400 pixel/blocks in total). However, if all these blocks are given the same value of `initial_populations=[N0]`, the simulation time goes back to ~200s.

I don't understand why the number of different nonlinear materials in the structure can affect the simulation time. In this model, the rate equation at each pixel doesn't involve any other pixels, meaning there is no extra d/dx term to calculate in the numerical iteration. Is there a way to fix or bypass this issue?

Thanks.

**Comment by oskooi** [comment, marked as answer]
> In that example, the simulation only takes ~200s. Then I divided the single two-level material block into 10 blocks with different `initial_populations`... the simulation time raised from ~200s to ~1440s. It can even go up to ~50000s if I define the material pixel by pixel(400 pixel/blocks in total). However, if all these blocks are given the same value of `initial_populations=[N0]`, the simulation time goes back to ~200s.

Note that each *unique* `Block` object of a multilevel-atomic susceptibility requires its own polarization fields and population-level densities to be stored and timestepped. The additional memory bandwidth necessary to read and write to these separate arrays will contribute to an overall longer runtime compared to a single `Block` object as the laser cavity.

cc @acerjan

**Reply by stevengj** [reply]
The polarization fields and population densities are **not per block**, they are per *material `medium`*.  You can have as many blocks or other objects as you want as long as they have an identical `medium` and it will share polarization arrays. 

That's why the time went down when you made the `initial_populations` equal.

In principle, I think we could probably share polarization arrays etcetera if the initial populations are different, as long as the other parameters are the same, since it just affects the initialization step.  We already do that if the `sigma` values are different, IIRC.
